File: src/game/progression.ts
/**
 * The campaign CONFIG: the bake-relevant pure definition of a single linear track of levels grouped
 * into **chapters** (each chapter adds a cumulative mechanic). Difficulty is **decoupled from board
 * size** (v2, see PLAN.md): a level's difficulty comes from where it sits on the per-chapter ease-in
 * curve (`targetPercentile`), NOT from its tube count. Board *shape* is just variety —
 * small/tall/medium/large all appear across a chapter, mixed within each difficulty band.
 *
 * Levels 1..N are **pre-baked** offline (`scripts/build-levels.ts` → `levels.data.ts`) by generating
 * a big pool across all shapes (the `SHAPES` menu), scoring each by a size-normalized difficulty
 * score, and assigning boards to the curve. Past the baked range (the plateau tail) and for the
 * endless mode, boards are generated live from the shape menu at high difficulty.
 *
 * Everything here feeds the bake's board OUTPUT, so the staleness guard (`scripts/levelVersion.ts`)
 * hashes this file. The runtime LOADING path (`getLevel`, the live generator, baked deserialization)
 * lives in `levelLoader.ts` so that tuning it does NOT force a re-bake.
 */
import { DEFAULT_CAPACITY } from './generator';
import type { HiddenGrid } from './hidden';
import type { Difficulty, GeneratedLevel, Mechanic, ParMode } from './types';

/** A board footprint the bake draws candidates from. `family` groups shapes for variety rotation. */
export interface Shape {
  family: 'small' | 'tall' | 'medium' | 'large';
  /** Number of tubes. */
  bottles: number;
  /** Number of distinct colors (each fills `capacity` cells). */
  colors: number;
  capacity: number;
}

/**
 * The shape menu (v2): the footprints the offline bake samples candidates from, spanning small,
 * tall (5-tube only, capacity swept up to 10 — a compact hard variation), medium, and large. This
 * is NOT a difficulty ladder — difficulty comes from the curve + per-board scoring; this is purely
 * the variety of board shapes that can appear at any difficulty. Every entry keeps `bottles - colors
 * >= 1` (at least one empty tube) so it's generatable. Capacity is capped at 10: a capacity-12 tube
 * gets too cramped and can clip/crowd neighbours on small screens, so it's served nowhere — baked or
 * live (see LIVE_SHAPES/RANDOM_SHAPES, which also enforce the cap).
 */
export const SHAPES: readonly Shape[] = [
  // Small classic (5 tubes, standard height).
  { family: 'small', bottles: 5, colors: 3, capacity: 4 },
  { family: 'small', bottles: 5, colors: 4, capacity: 4 },
  // Small TALL (5 tubes only): few tubes, very dense — hard in a compact footprint.
  { family: 'tall', bottles: 5, colors: 3, capacity: 6 },
  { family: 'tall', bottles: 5, colors: 4, capacity: 6 },
  { family: 'tall', bottles: 5, colors: 4, capacity: 8 },
  { family: 'tall', bottles: 5, colors: 4, capacity: 10 },
  // Medium. Tube counts are restricted to multiples of the 5-per-row grid (5/10/15) so every board
  // fills whole rows — 9- or 13-tube boards leave a ragged last row on mobile.
  { family: 'medium', bottles: 10, colors: 7, capacity: 4 },
  { family: 'medium', bottles: 10, colors: 8, capacity: 4 },
  // Large.
  { family: 'large', bottles: 15, colors: 11, capacity: 4 },
  { family: 'large', bottles: 15, colors: 12, capacity: 4 },
];

/**
 * Shapes the live path (plateau tail + endless mode + un-baked fallback) draws from — hard-leaning.
 * Tall tubes are capped at capacity 10 (the baked campaign's max height): capacity-12 tubes get too
 * cramped and can clip or crowd neighbours on small screens, so the live path never serves them.
 */
const LIVE_SHAPES: readonly Shape[] = SHAPES.filter(
  (s) => s.family === 'large' || (s.family === 'tall' && s.capacity >= 8 && s.capacity <= 10),
);

/** Number of levels per chapter. Levels 1..CHAPTER_LEN are chapter 0, etc. */
export const CHAPTER_LEN = 60;

/**
 * Cumulative mechanic sets, indexed by chapter. Chapter 0 is the base game; chapter 1 adds the
 * hidden-colors mechanic. Past the last defined chapter, play plateaus in the final chapter.
 */
const MECHANIC_SETS: readonly (readonly Mechanic[])[] = [
  [], // chapter 0 — base game
  ['hidden'], // chapter 1 — + hidden colors
];

/** Highest chapter we actually have content for. */
export const DEFINED_CHAPTERS = MECHANIC_SETS.length;

/** Total number of campaign levels the offline bake produces (every defined chapter, full length). */
export const CAMPAIGN_LENGTH = DEFINED_CHAPTERS * CHAPTER_LEN;

/** The chapter a level belongs to (clamped to the last defined chapter — the plateau). */
export function chapterForLevel(level: number): number {
  const idx = Math.max(0, Math.floor(level) - 1);
  return Math.min(Math.floor(idx / CHAPTER_LEN), DEFINED_CHAPTERS - 1);
}

/** The mechanics active at a level (its chapter's cumulative set). */
export function mechanicsForLevel(level: number): readonly Mechanic[] {
  return MECHANIC_SETS[chapterForLevel(level)]!;
}

/** Chapter index and within-chapter position (0-based), with plateau clamping past defined chapters. */
function chapterPos(level: number): { chapter: number; pos: number } {
  const idx = Math.max(0, Math.floor(level) - 1);
  const rawChapter = Math.floor(idx / CHAPTER_LEN);
  if (rawChapter >= DEFINED_CHAPTERS) {
    return { chapter: DEFINED_CHAPTERS - 1, pos: CHAPTER_LEN - 1 }; // plateau at the chapter's end
  }
  return { chapter: rawChapter, pos: idx % CHAPTER_LEN };
}

/**
 * xmur3 string hash -> a 32-bit seed. Mixing the level number (rather than using it raw)
 * decorrelates adjacent levels so consecutive boards share no structure.
 */
function xmur3(str: string): number {
  let h = 1779033703 ^ str.length;
  for (let i = 0; i < str.length; i++) {
    h = Math.imul(h ^ str.charCodeAt(i), 3432918353);
    h = (h << 13) | (h >>> 19);
  }
  h = Math.imul(h ^ (h >>> 16), 2246822507);
  h = Math.imul(h ^ (h >>> 13), 3266489909);
  h ^= h >>> 16;
  return h >>> 0;
}

/** Deterministic seed for a level. `salt` is bumped only when generation fails (see below). */
export function seedForLevel(level: number, salt = 0): number {
  return xmur3(`magic-color:L${level}:${salt}`);
}

/** Par floor for a level: reject trivial boards, ramping mildly across the chapter. */
function parFloorFor(colors: number, posInChapter: number): number {
  const base = Math.round(colors * 1.3);
  const ramp = Math.round((posInChapter / CHAPTER_LEN) * colors * 0.8);
  return base + ramp;
}

/**
 * Offline-bake difficulty curve (consumed by `scripts/build-levels.ts`, not the runtime).
 *
 * Returns where a level should sit on the difficulty curve as a PERCENTILE (0..1) of the chapter's
 * scored candidate pool — a self-calibrating target, so we don't predict absolute difficulty. The
 * curve EASES IN within each chapter (gentle openers) and starts from a HIGHER FLOOR in later
 * chapters, so even the early levels of the hidden chapter are harder than the base chapter's. The
 * three knobs below are the dials (see PLAN.md). Past the last defined chapter the percentile
 * plateaus.
 */
const CURVE = {
  /** Percentile of the very first level (chapter 0, position 0). */
  baseFloor: 0.15,
  /** How much higher each subsequent chapter's floor (and thus its whole curve) sits. */
  chapterFloorStep: 0.12,
  /** How far the percentile climbs from a chapter's start to its end. */
  span: 0.7,
  /** Ease-in exponent: >1 makes the opening levels ramp gently and the back half steeper. */
  easeExp: 1.6,
} as const;

export function targetPercentile(level: number): number {
  const { chapter, pos } = chapterPos(level);
  const t = CHAPTER_LEN <= 1 ? 0 : pos / (CHAPTER_LEN - 1);
  const eased = Math.pow(t, CURVE.easeExp);
  const p = CURVE.baseFloor + chapter * CURVE.chapterFloorStep + eased * CURVE.span;
  return Math.min(1, Math.max(0, p));
}

/**
 * The level's difficulty label — one of three buckets derived from its position on the curve, NOT
 * from tube count (v2). A pure function of the level, so the bake and the live path agree.
 */
export function phaseForLevel(level: number): Difficulty {
  const p = targetPercentile(level);
  if (p < 1 / 3) return 'easy';
  if (p < 2 / 3) return 'normal';
  return 'hard';
}

/** The recipe for generating a level LIVE: a footprint plus chapter/phase, seed, and par knobs. */
export interface LevelPlan {
  level: number;
  chapter: number;
  phase: Difficulty;
  colors: number;
  bottles: number;
  capacity: number;
  seed: number;
  minPar: number;
  parMode: ParMode;
  mechanics: readonly Mechanic[];
}

/** A generated, playable level annotated with its campaign metadata. */
export interface PlayableLevel extends GeneratedLevel {
  level: number;
  chapter: number;
  phase: Difficulty;
  mechanics: readonly Mechanic[];
  /** Initial concealment overlay (all-false unless this chapter has the `hidden` mechanic). */
  hidden: HiddenGrid;
  /**
   * Achievable near-optimal move count (the solution replayed under the real capped/reveal
   * rules). Basis for star thresholds — see `stars.ts`. Differs from `minMoves` (bulk solution
   * length) on hidden levels, where capping forces more, smaller pours.
   */
  optimal: number;
  /** 2★ ceiling: the adjusted near-optimal band's upper bound (always `> optimal`). See `stars.ts`. */
  twoStarMax: number;
}

/**
 * Recipe for generating a level LIVE (the plateau tail past the baked range, and the safety
 * fallback for any un-baked level). Baked levels do NOT go through this — their board comes straight
 * from `levels.data.ts`. The live path picks a hard-leaning shape (rotated by level for variety),
 * since live levels only occur at/after the plateau where play sits at Hard.
 */
export function planForLevel(level: number): LevelPlan {
  const { chapter, pos } = chapterPos(level);
  const shape = LIVE_SHAPES[Math.max(0, Math.floor(level) - 1) % LIVE_SHAPES.length]!;

  return {
    level,
    chapter,
    phase: phaseForLevel(level),
    colors: shape.colors,
    bottles: shape.bottles,
    capacity: shape.capacity,
    seed: seedForLevel(level),
    minPar: parFloorFor(shape.colors, pos),
    // Exact par only for small, standard-height boards; exact BFS explodes on bigger or taller ones.
    parMode: shape.bottles <= 6 && shape.capacity <= DEFAULT_CAPACITY ? 'optimal' : 'proxy',
    mechanics: MECHANIC_SETS[chapter]!,
  };
}

File: core/src/progression.rs
//! Campaign config — port of `src/game/progression.ts`: the shape menu, chapters, mechanic
//! densities, level seeds (xmur3), the par floor, and the difficulty curve. These constants
//! are shared with JS and G5-checked; `seedForLevel` is exact (integer hash), while
//! `targetPercentile` goes through `pow` and is only tolerance-comparable cross-language
//! (deterministic across our own targets via libm — see `jsnum`).

use crate::generator::DEFAULT_CAPACITY;
use crate::jsnum::{js_round, pow};

#[derive(Clone, Copy, PartialEq, Eq, Debug)]
pub enum Family {
    Small,
    Tall,
    Medium,
    Large,
}

impl Family {
    pub fn name(self) -> &'static str {
        match self {
            Family::Small => "small",
            Family::Tall => "tall",
            Family::Medium => "medium",
            Family::Large => "large",
        }
    }
}

#[derive(Clone, Copy, Debug)]
pub struct Shape {
    pub family: Family,
    pub bottles: usize,
    pub colors: usize,
    pub capacity: u8,
}

/// The shape menu — must stay in the exact order of `SHAPES` in `progression.ts` (bake pool
/// assembly order depends on it).
pub const SHAPES: [Shape; 10] = [
    Shape { family: Family::Small, bottles: 5, colors: 3, capacity: 4 },
    Shape { family: Family::Small, bottles: 5, colors: 4, capacity: 4 },
    Shape { family: Family::Tall, bottles: 5, colors: 3, capacity: 6 },
    Shape { family: Family::Tall, bottles: 5, colors: 4, capacity: 6 },
    Shape { family: Family::Tall, bottles: 5, colors: 4, capacity: 8 },
    Shape { family: Family::Tall, bottles: 5, colors: 4, capacity: 10 },
    Shape { family: Family::Medium, bottles: 10, colors: 7, capacity: 4 },
    Shape { family: Family::Medium, bottles: 10, colors: 8, capacity: 4 },
    Shape { family: Family::Large, bottles: 15, colors: 11, capacity: 4 },
    Shape { family: Family::Large, bottles: 15, colors: 12, capacity: 4 },
];

pub const CHAPTER_LEN: usize = 60;

#[derive(Clone, Copy, PartialEq, Eq, Debug)]
pub enum Mechanic {
    Hidden,
    Funnel,
    Ice,
}

/// Cumulative mechanic sets by chapter (chapter 0 = base game).
pub const MECHANIC_SETS: [&[Mechanic]; 4] = [
    &[],
    &[Mechanic::Hidden],
    &[Mechanic::Hidden, Mechanic::Funnel],
    &[Mechanic::Hidden, Mechanic::Funnel, Mechanic::Ice],
];

pub const DEFINED_CHAPTERS: usize = MECHANIC_SETS.len();
pub const CAMPAIGN_LENGTH: usize = DEFINED_CHAPTERS * CHAPTER_LEN;

pub fn chapter_for_level(level: usize) -> usize {
    let idx = level.saturating_sub(1);
    (idx / CHAPTER_LEN).min(DEFINED_CHAPTERS - 1)
}

pub fn mechanics_for_level(level: usize) -> &'static [Mechanic] {
    MECHANIC_SETS[chapter_for_level(level)]
}

/// Per-mechanic application density (the `prob` fed to each compute*).
#[derive(Clone, Copy, Debug, PartialEq)]
pub struct MechanicDensity {
    pub hidden: f64,
    pub funnel: f64,
    pub ice: f64,
}

const SIGNATURE_DENSITY: MechanicDensity = MechanicDensity { hidden: 0.7, funnel: 0.62, ice: 0.8 };
const INHERITED_DENSITY: MechanicDensity = MechanicDensity { hidden: 0.15, funnel: 0.3, ice: 0.3 };
const BALANCED_DENSITY: MechanicDensity = MechanicDensity { hidden: 0.3, funnel: 0.5, ice: 0.5 };

/// The mechanic a chapter INTRODUCES (last in its cumulative set).
pub fn signature_mechanic(chapter: usize) -> Option<Mechanic> {
    MECHANIC_SETS[chapter.min(DEFINED_CHAPTERS - 1)].last().copied()
}

pub fn campaign_density(chapter: usize) -> MechanicDensity {
    let sig = signature_mechanic(chapter);
    MechanicDensity {
        hidden: if sig == Some(Mechanic::Hidden) { SIGNATURE_DENSITY.hidden } else { INHERITED_DENSITY.hidden },
        funnel: if sig == Some(Mechanic::Funnel) { SIGNATURE_DENSITY.funnel } else { INHERITED_DENSITY.funnel },
        ice: if sig == Some(Mechanic::Ice) { SIGNATURE_DENSITY.ice } else { INHERITED_DENSITY.ice },
    }
}

pub fn balanced_density() -> MechanicDensity {
    BALANCED_DENSITY
}

/// Chapter index and 0-based within-chapter position, plateau-clamped.
pub fn chapter_pos(level: usize) -> (usize, usize) {
    let idx = level.saturating_sub(1);
    let raw = idx / CHAPTER_LEN;
    if raw >= DEFINED_CHAPTERS {
        (DEFINED_CHAPTERS - 1, CHAPTER_LEN - 1)
    } else {
        (raw, idx % CHAPTER_LEN)
    }
}

/// xmur3 string hash → 32-bit seed — bit-exact port (JS `Math.imul` / rotate semantics).
fn xmur3(s: &str) -> u32 {
    let bytes = s.as_bytes();
    let mut h: u32 = 1779033703 ^ bytes.len() as u32;
    for &b in bytes {
        h = (h ^ b as u32).wrapping_mul(3432918353);
        h = h.rotate_left(13);
    }
    h = (h ^ (h >> 16)).wrapping_mul(2246822507);
    h = (h ^ (h >> 13)).wrapping_mul(3266489909);
    h ^ (h >> 16)
}

/// Deterministic seed for a level (`seedForLevel`).
pub fn seed_for_level(level: usize, salt: u32) -> u32 {
    xmur3(&format!("magic-color:L{level}:{salt}"))
}

/// Par floor: reject trivial boards, ramping mildly across the chapter (`parFloorFor`).
pub fn par_floor_for(colors: usize, pos_in_chapter: usize) -> u32 {
    let base = js_round(colors as f64 * 1.3);
    let ramp = js_round((pos_in_chapter as f64 / CHAPTER_LEN as f64) * colors as f64 * 0.8);
    (base + ramp) as u32
}

/// Difficulty-curve knobs (`CURVE`).
const BASE_FLOOR: f64 = 0.15;
const CHAPTER_FLOOR_STEP: f64 = 0.12;
const SPAN: f64 = 0.7;
const EASE_EXP: f64 = 1.6;

/// Where a level should sit on its chapter's score distribution, as a percentile in [0, 1].
/// Goes through `pow` — tolerance-compare cross-language (see module docs).
pub fn target_percentile(level: usize) -> f64 {
    let (chapter, pos) = chapter_pos(level);
    let t = if CHAPTER_LEN <= 1 { 0.0 } else { pos as f64 / (CHAPTER_LEN - 1) as f64 };
    let eased = pow(t, EASE_EXP);
    let p = BASE_FLOOR + chapter as f64 * CHAPTER_FLOOR_STEP + eased * SPAN;
    p.clamp(0.0, 1.0)
}

#[derive(Clone, Copy, PartialEq, Eq, Debug)]
pub enum Phase {
    Easy,
    Normal,
    Hard,
}

impl Phase {
    pub fn name(self) -> &'static str {
        match self {
            Phase::Easy => "easy",
            Phase::Normal => "normal",
            Phase::Hard => "hard",
        }
    }
}

pub fn phase_for_level(level: usize) -> Phase {
    let p = target_percentile(level);
    if p < 1.0 / 3.0 {
        Phase::Easy
    } else if p < 2.0 / 3.0 {
        Phase::Normal
    } else {
        Phase::Hard
    }
}

/// Whether the bake measures exact (BFS) par for a shape (`planForLevel`'s parMode rule).
pub fn exact_par_shape(bottles: usize, capacity: u8) -> bool {
    bottles <= 6 && capacity <= DEFAULT_CAPACITY
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn chapters_and_plateau() {
        assert_eq!(chapter_for_level(1), 0);
        assert_eq!(chapter_for_level(60), 0);
        assert_eq!(chapter_for_level(61), 1);
        assert_eq!(chapter_for_level(240), 3);
        assert_eq!(chapter_for_level(999), 3);
        assert_eq!(chapter_pos(999), (3, 59)); // plateau clamps to the chapter's end
    }

    #[test]
    fn densities_follow_signature_role() {
        assert_eq!(campaign_density(1).hidden, 0.7); // hidden is chapter 1's signature
        assert_eq!(campaign_density(2).hidden, 0.15); // inherited thereafter
        assert_eq!(campaign_density(2).funnel, 0.62);
        assert_eq!(campaign_density(3).ice, 0.8);
    }

    #[test]
    fn curve_is_monotone_within_a_chapter() {
        for lvl in 1..60 {
            assert!(target_percentile(lvl + 1) >= target_percentile(lvl));
        }
        // Later chapters start above earlier ones' floors.
        assert!(target_percentile(61) > target_percentile(1));
    }
}
